FirebaseAuth/Sources/Swift/Backend/AuthBackend.swift

```diff
@@ -0,0 +1,386 @@
+  override init() {
+    RPCIssuer = AuthBackendRPCIssuerImplementation()
+  }
+
+  /** @fn postWithRequest:response:callback:
+      @brief Calls the RPC using HTTP POST.
+      @remarks Possible error responses:
+          @see FIRAuthInternalErrorCodeRPCRequestEncodingError
+          @see FIRAuthInternalErrorCodeJSONSerializationError
+          @see FIRAuthInternalErrorCodeNetworkError
+          @see FIRAuthInternalErrorCodeUnexpectedErrorResponse
+          @see FIRAuthInternalErrorCodeUnexpectedResponse
+          @see FIRAuthInternalErrorCodeRPCResponseDecodingError
+      @param request The request.
+      @param response The empty response to be filled.
+      @param callback The callback for both success and failure.
+   */
+  fileprivate func post(withRequest request: AuthRPCRequest,
+                        callback: @escaping ((AuthRPCResponse?, Error?) -> Void)) {
+    let response = request.response
+    post(withRequest: request, response: response) { error in
+      if let error = error {
+        callback(nil, error)
+      } else if let mfaError = AuthBackendRPCImplementation.generateMFAError(response: response) {
+        callback(nil, mfaError)
+      } else {
+        callback(response, nil)
+      }
+    }
+  }
+
+  #if os(iOS)
+    private class func generateMFAError(response: AuthRPCResponse) -> Error? {
+      if let mfaResponse = response as? EmailLinkSignInResponse,
+         mfaResponse.IDToken == nil,
+         let enrollments = mfaResponse.MFAInfo {
+        var info: [MultiFactorInfo] = []
+        for enrollment in enrollments {
+          info.append(MultiFactorInfo(proto: enrollment))
+        }
+        return AuthErrorUtils.secondFactorRequiredError(
+          pendingCredential: mfaResponse.MFAPendingCredential,
+          hints: info
+        )
+      } else {
+        return nil
+      }
+    }
+  #else
+    private class func generateMFAError(response: AuthRPCResponse) -> Error? {
+      return nil
+    }
+  #endif
+
+  /** @fn postWithRequest:response:callback:
+      @brief Calls the RPC using HTTP POST.
+      @remarks Possible error responses:
+          @see FIRAuthInternalErrorCodeRPCRequestEncodingError
+          @see FIRAuthInternalErrorCodeJSONSerializationError
+          @see FIRAuthInternalErrorCodeNetworkError
+          @see FIRAuthInternalErrorCodeUnexpectedErrorResponse
+          @see FIRAuthInternalErrorCodeUnexpectedResponse
+          @see FIRAuthInternalErrorCodeRPCResponseDecodingError
+      @param request The request.
+      @param response The empty response to be filled.
+      @param callback The callback for both success and failure.
+   */
+  fileprivate func post(withRequest request: AuthRPCRequest,
+                        response: AuthRPCResponse,
+                        callback: @escaping ((Error?) -> Void)) {
+    var bodyData: Data?
+    if request.containsPostBody?() != nil {
+      do {
+        // TODO: Can unencodedHTTPRequestBody ever throw?
```

Comment:
They don't today, but there are a few fatalErrors that might better be implemented as thrown errors.. Although perhaps the case of 'containsPostBody' returning false could perhaps be modelled differently so that the failing unencodedHTTPRequestBody could only be called when a body exists...